src/vmm/src/device_manager/mmio.rs

```diff
@@ -76,12 +78,50 @@ pub struct MMIODeviceInfo {
     pub irqs: Vec<u32>,
 }
 
+impl Aml for MMIODeviceInfo {
+    fn append_aml_bytes(&self, v: &mut Vec<u8>) {
+        let dev_id = self.irqs[0] - arch::IRQ_BASE;
+        debug!(
+            "acpi: Building AML for VirtIO device _SB_.V{:03}. memory range: {:#010x}:{} irq: {}",
+            dev_id, self.addr, self.len, self.irqs[0]
+        );
+        aml::Device::new(
+            format!("V{:03}", dev_id).as_str().into(),
+            vec![
+                &aml::Name::new("_HID".into(), &"LNRO0005"),
+                &aml::Name::new("_UID".into(), &dev_id),
+                &aml::Name::new("_CCA".into(), &aml::ONE),
+                &aml::Name::new(
+                    "_CRS".into(),
+                    &aml::ResourceTemplate::new(vec![
+                        &aml::Memory32Fixed::new(
+                            true,
+                            self.addr.try_into().unwrap(),
+                            self.len.try_into().unwrap(),
+                        ),
+                        &aml::Interrupt::new(true, true, false, false, self.irqs[0]),
+                    ]),
+                ),
+            ],
+        )
+        .append_aml_bytes(v);
+    }
+}
+
 /// Manages the complexities of registering a MMIO device.
 #[derive(Debug)]
 pub struct MMIODeviceManager {
     pub(crate) bus: crate::devices::Bus,
     pub(crate) resource_allocator: Rc<ResourceAllocator>,
     pub(crate) id_to_dev_info: HashMap<(DeviceType, String), MMIODeviceInfo>,
+    // We create the AML byte code for every VirtIO device in the order we build
+    // it, so that we ensure the root block device is appears first in the DSDT.
+    // This is needed, so that the root device appears as `/dev/sda` in the guest
+    // filesystem.
+    // The alternative would be that we iterate the bus to get the data after all
+    // of the devices are build. However, iterating the bus won't give us the
+    // devices in the order they were added.
+    pub(crate) dsdt_data: Vec<u8>,
```

Comment:
We already store `MMIODeviceInfo` structs in the hash map of the MMIODeviceManager. Keeping them her would mean that we need to clone them. What's the added value of do so?

Reply:
The idea here is to store MMIO related info in the `MMIOManager`. By storing ACPI info here we mix things that should not know about each other.

Reply:
> that should not know about each other

Why? And why would it be better to store the same object twice?

Reply:
Because MMIO object will only have MMIO info, not ACPI info.

Reply:
> The ACPI info is MMIO info.

Isn't it the way around? We convert `MMIODeviceInfo` into ACPI aml bytes.

Reply:
Yes, we 're saying the same thing. `dsdt_data` here is the `MMIODeviceInfo` translated into AML bytecode, it is not "random ACPI data". The reason why we store it there is described in a previous comment. It's an (unfortunate) way for us to keep this info stored in the order it should appear in the guest. Otherwise, I would just create them on the fly iterating the hash table.

I don't see any reason why we should store a copy of `MMIODeviceInfo`.

Reply:
Ok, so this can be removed once `MMIODeviceManager` will be able to remember the order of device insertions. Let's create a TODO/issue to remember this.